qiskit_machine_learning/neural_networks/sampling_qnn.py

```diff
@@ -0,0 +1,177 @@
+
+from .sampler_qnn import SamplerQNN
+
+
+class SamplingQNN(SamplerQNN):
+    """
+    A sampling neural network that generate samples instead of probabilities. This network extends
+    :class:`~qiskit_machine_learning.neural_networks.SamplerQNN`, generates samples out of the
+    quasi-probabilities returned by the sampler primitive, and returns them. This network to be
+    replaced/updated when a new primitive that can return samples will be introduced.
+
+    This network provides only the ``forward`` method. The method constructs a multinomial
+    distribution for each quasi-distribution from the sampler results, samples ``num_shots`` values
+    from this distribution, and applies ``interpret`` to the sampled values. The interpreted values
+    are returned in an array of shape ``(num_samples, num_shots, *interpret_shape)``. Where
+    ``num_samples`` is the number of samples (or batch size in other words) in the dataset and
+    ``interpret_shape`` is the shape of the output of the ``interpret`` function. If the function
+    returns a single number then the interpret shape is ``1``, if the function returns an array,
+    then the interpret shape is the shape of this array.
+
+    The ``backward`` always returns a tuple ``(None, None)``.
+
+    In this example the network generates samples without custom interpretation. The circuit has two
+    qubits, thus there are four possible outcomes, but in this setup the network generates evenly
+    distributed outcomes of ``0`` and ``1``. The network does not take any input data.
+
+    .. code-block::
+
+        import numpy as np
+        from qiskit import QuantumCircuit
+        from qiskit.circuit import Parameter
+
+        from qiskit_machine_learning.neural_networks.sampling_qnn import SamplingQNN
+
+        qc = QuantumCircuit(2)
+
+        qc.ry(Parameter("0"), 0)
+        qc.ry(Parameter("1"), 1)
+        qc.rx(np.pi / 2, 0)
+
+        qnn = SamplingQNN(circuit=qc, num_shots=10, input_params=[], weight_params=qc.parameters)
+        qnn.forward(input_data=None, weights=[0, 0])
+    """
+
+    def __init__(
+        self,
+        *,
+        circuit: QuantumCircuit,
+        num_shots: int | None = None,
+        sampler: BaseSampler | None = None,
+        input_params: Sequence[Parameter] | None = None,
+        weight_params: Sequence[Parameter] | None = None,
+        interpret: Callable[[int], int | tuple[int, ...]] | None = None,
+    ):
+        """
+        Args:
+            circuit: The parametrized quantum circuit that generates the samples of this network.
+            sampler: The sampler primitive used to compute the neural network's results.
+                If ``None`` is given, a default instance of the reference sampler defined
+                by :class:`~qiskit.primitives.Sampler` will be used.
+            num_shots: A positive number of shots for each circuit to sample from the
+                quasi-distribution generated by the sampler. The default is 1024.
```

Comment:
Since a Sampler, which this takes, has shots would it be helpful to explain the difference to this, and how would I normally set them say wrt to one another if I need to care about this.   

I wonder whether SamplerQNN and SamplingQNN are too similar in name and will lead to confusion - having said that I have no idea for anything different for the latter.

Reply:
This is tricky. In general, they both have the same meaning, but they are interpreted differently. Shots in the Sampler is used to generate distributions. In the QNN shots (or `num_shots`) is used to generate samples (circuit measurements/outcomes?) from the generated distributions. Basically, the QNN does the reverse operation of the Sampler since we don't have direct access to the circuit execution results. The problem is that `shots` has a proper meaning in quantum computing. Same for `samples`, it has a proper meaning of the number of samples in a dataset. In fact, the QNN can re-use the value set to `shots` if present. So, it is confusing, I agree. I'm happy to rename `num_shots` to somewhat else, but I have no idea too. And I don't like the name of the QNN as well.

Of course, I can document the difference between these two parameters/meanings in more details. Any ideas on the names are really welcome.

Reply:
For naming of this parameter, `num_samples` might be nice.
BTW, since `QuasiDistribution` sometimes has `shots`. It would be nice to have the option of using this `shots` to generate the sequence of outputs. In this case, I would suggest shuffling outcomes instead of bootstrap (resampling).